rs/boundary_node/salt_sharing/canister/canister.rs

```diff
@@ -0,0 +1,161 @@
+        .await
+        .map_err(|err| {
+            format!(
+                "call to raw_rand of management canister failed: rejection_code={:?}, message={:?}",
+                err.0, err.1
+            )
+        })?;
+    RNG.with_borrow_mut(|rng| *rng = Some(ChaCha20Rng::from_seed(seed)));
+    Ok(())
+}
+
+// Schedules the seeding of randomness with an optional recurring period.
+fn schedule_rng_seeding(delay: Duration, period: Option<Duration>) {
+    set_timer(delay, move || {
+        spawn(async move {
+            if let Err(err) = seed_randomness().await {
+                // TODO: substitute with logs
+                ic_cdk::println!("Failed to seed randomness {err}");
+            }
+            // Schedule next seeding
+            if let Some(period) = period {
+                schedule_rng_seeding(period, Some(period));
+            }
+        })
+    });
+}
+
+// Sets an execution timer (delayed future task) and returns immediately.
+fn schedule_salt_generation(strategy: SaltGenerationStrategy) {
+    match strategy {
+        SaltGenerationStrategy::StartOfMonth => {
+            // Add a small delay to ensure execution starts after the month started
+            let delay = delay_till_next_month(time()) + START_OF_MONTH_DELAY;
+            set_timer(delay, || {
+                if let Err(err) = regenerate_salt() {
+                    // TODO: substitute with logs
+                    ic_cdk::println!("Failed to regenerate salt {err}");
+                }
+                // Function is called recursively to schedule next execution
+                schedule_salt_generation(strategy);
+            });
+        }
+    }
+}
+
+fn is_salt_init() -> bool {
+    SALT.with(|cell| cell.borrow().get(&())).is_some()
+        && LAST_SALT_NS.with(|cell| cell.borrow().get(&())).is_some()
+}
+
+fn get_salt_response() -> Result<SaltResponse, GetSaltError> {
+    let salt = SALT
+        .with(|cell| cell.borrow().get(&()))
+        .ok_or(GetSaltError::SaltNotInitialized)?;
+    // TODO: use ms or secs.
+    let salt_id = LAST_SALT_NS
+        .with(|cell| cell.borrow_mut().insert((), time()))
+        .ok_or(GetSaltError::SaltNotInitialized)?;
+    Ok(SaltResponse { salt, salt_id })
+}
+
+fn seed_rng_and_regenerate_salt(remaining_attempts: i32, delay: Duration) {
+    if remaining_attempts <= 0 {
+        return;
+    }
+    set_timer(delay, move || {
+        spawn(async move { try_seed_and_regenerate_salt(remaining_attempts).await });
+    });
+}
+
+// Attempts to initialize random seed and generate the salt
+async fn try_seed_and_regenerate_salt(remaining_attempts: i32) {
+    if let Err(err) = seed_randomness().await {
+        // TODO: substitute with logs
+        ic_cdk::println!("Failed to seed randomness: {err}. Retrying...");
+        seed_rng_and_regenerate_salt(remaining_attempts - 1, SEED_INIT_RETRY_INTERVAL);
+        return; // Stop further execution to avoid redundant retries
```

Comment:
I didn't get the comment: by returning here you avoid the subsequent if-statement, but I don't see what retries you refer to.

Reply:
`fn try_seed_and_regenerate_salt()` calls a retry function `fn seed_rng_and_regenerate_salt` in potentially two places; if `seed_randomness()` fails or if `regenerate_salt()` fails. So `return` would guarantee  that [this subsequent retry](https://github.com/dfinity/ic/pull/3650) is not called.

Reply:
Now I see what you mean, but I'm not sure if this is apparent from the comment without further context: e.g., it is not clear that regenerate_salt is guaranteed to fail and trigger one more retry.